File: IPRecord.cpp
#include "IPRecord.hpp"

// constructor sets starting values for a new ip record.
// risk score and attempt count start at zero since no connections have happened yet.
IPRecord::IPRecord(std::string ip) {
    ipAddress = ip;
    riskScore = 0;
    attemptCount = 0;
    historyHead = nullptr;
}

// destructor walks the linked list and deletes every node.
// this prevents memory leaks since each node was allocated with new.
IPRecord::~IPRecord() {
    ConnectionRecord* current = historyHead;
    while (current != nullptr) {
        ConnectionRecord* temp = current;
        current = current->next;
        delete temp;
    }
}

// this function adds a new connection attempt to the front of the history list.
// adding to the front keeps insertion at constant time instead of walking the list.
void IPRecord::addConnection(std::string timestamp, int port, bool flagged) {
    ConnectionRecord* newRecord = new ConnectionRecord(timestamp, port, flagged);
    newRecord->next = historyHead;
    historyHead = newRecord;
    attemptCount++;
}

// this function raises the risk score by the given amount.
// used whenever a connection looks suspicious or repeated too often.
void IPRecord::increaseRisk(int amount) {
    riskScore += amount;
}

// getters below just return the private data so other classes can read it.
std::string IPRecord::getIP() const {
    return ipAddress;
}

int IPRecord::getRiskScore() const {
    return riskScore;
}

int IPRecord::getAttemptCount() const {
    return attemptCount;
}

ConnectionRecord* IPRecord::getHistory() const {
    return historyHead;
}

File: web/src/models/IPRecord.ts
// this class represents one single connection attempt from an ip.
// this is the entry used in the connection history array.
export interface ConnectionRecord {
  timestamp: string;
  port: number;
  flagged: boolean;
}

// this class holds all the info tracked for a single ip address.
// it stores the ip itself, a risk score, and a history of its connections.
export class IPRecord {
  private ipAddress: string;
  private riskScore: number;
  private attemptCount: number;
  private history: ConnectionRecord[];

  // constructor sets up a new ip record with starting values.
  constructor(ip: string) {
    this.ipAddress = ip;
    this.riskScore = 0;
    this.attemptCount = 0;
    this.history = [];
  }

  // this function adds a new connection attempt to the history array.
  // adding to the front keeps the newest connection easy to find first.
  addConnection(timestamp: string, port: number, flagged: boolean): void {
    const record: ConnectionRecord = { timestamp, port, flagged };
    this.history.unshift(record);
    this.attemptCount++;
  }

  // this function raises the risk score by the given amount.
  // used whenever a connection looks suspicious or repeated too often.
  increaseRisk(amount: number): void {
    this.riskScore += amount;
  }

  // getters below just return the private data so other classes can read it.
  getIP(): string {
    return this.ipAddress;
  }

  getRiskScore(): number {
    return this.riskScore;
  }

  getAttemptCount(): number {
    return this.attemptCount;
  }

  getHistory(): ConnectionRecord[] {
    return this.history;
  }
}